aws-lc-rs/src/key_transport.rs

```diff
@@ -0,0 +1,408 @@
+            }
+            kdf(&shared_secret)
+        }
+    }
+
+    /// Creates a new KEM private key from raw bytes. This method is NOT meant to generate
+    /// a new private key, rather it restores a `KemPrivateKey` that was previously converted
+    /// to raw bytes.
+    ///
+    /// `alg` is the `KemAlgorithm` to be associated with the generated `KemPrivateKey`
+    ///
+    /// `bytes` is a slice of raw bytes representing a `KemPrivateKey`
+    ///
+    /// # Errors
+    /// `error::KeyRejected` when operation fails during key creation.
+    ///
+    pub fn new(alg: &'static KemAlgorithm, bytes: &[u8]) -> Result<Self, KeyRejected> {
+        if alg.secret_key_size() != bytes.len() {
+            return Err(KeyRejected::unexpected_error());
+        }
+        unsafe {
+            let privkey = LcPtr::new(EVP_PKEY_kem_new_raw_secret_key(
+                alg.id.nid(),
+                bytes.as_ptr(),
+                bytes.len(),
+            ))?;
+            Ok(KemPrivateKey {
+                algorithm: alg,
+                pkey: privkey,
+                priv_key: bytes.to_owned().into(),
+            })
+        }
+    }
+}
+
+impl Drop for KemPrivateKey {
+    fn drop(&mut self) {
+        self.priv_key.zeroize();
+    }
+}
+
+impl AsRef<[u8]> for KemPrivateKey {
+    fn as_ref(&self) -> &[u8] {
+        &self.priv_key
+    }
+}
+
+/// A serializable public key usable with KEMS. This can be constructed
+/// from a `KemPrivateKey` or constructed from raw bytes.
+#[derive(Debug)]
+pub struct KemPublicKey {
+    algorithm: &'static KemAlgorithm,
+    pkey: LcPtr<*mut EVP_PKEY>,
+    pub_key: Box<[u8]>,
+}
+
+impl KemPublicKey {
+    /// Return the algorithm associated with the given KEM public key.
+    #[must_use]
+    pub fn algorithm(&self) -> &'static KemAlgorithm {
+        self.algorithm
+    }
+
+    /// Performs the encapsulate operation using the current KEM public key, generating a ciphertext
+    /// and associated shared secret.
+    ///
+    /// After the encapsulation is finished, `encapsulate` calls `kdf` with the ciphertext and raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    ///
+    pub fn encapsulate<F, R>(&self, kdf: F) -> Result<R, Unspecified>
+    where
+        F: FnOnce(&[u8], &[u8]) -> Result<R, Unspecified>,
+    {
+        unsafe {
+            let ctx = LcPtr::new(EVP_PKEY_CTX_new(*self.pkey, null_mut()))?;
+
+            let mut ciphertext: Vec<u8> = vec![0u8; self.algorithm.cipher_text_size()];
+            let mut shared_secret: Vec<u8> = vec![0u8; self.algorithm.shared_secret_size()];
+
+            if EVP_PKEY_encapsulate(
+                *ctx,
+                ciphertext.as_mut_ptr(),
+                &mut self.algorithm.cipher_text_size(),
+                shared_secret.as_mut_ptr(),
+                &mut self.algorithm.shared_secret_size(),
```

Comment:
Don't pass mutable pointers to return values:
```
    let mut cipher_text_size = self.algorithm.cipher_text_size();
    let mut shared_secret_size = self.algorithm.shared_secret_size();
    if EVP_PKEY_encapsulate(
        *ctx,
        ciphertext.as_mut_ptr(),
        &mut cipher_text_size,
        shared_secret.as_mut_ptr(),
        &mut shared_secret_size,
    ) != 1
    {
        return Err(Unspecified);
    }
```

As above, do we need to check that these values were not changed?

Reply:
We shouldn't need to check that these are unchanged because we know that our getters will return the correct length